### ISSUE TITLE
App has stopped functioning

### ISSUE THREAD

**Issue by Boballen55** [nextcloud/cookbook#734]
**Description**
Cookbook has stopped loading. 

**Reproduction**
1. Go to Cookbook app

**Expected behavior**
Cookbook recipes to load

**Actual behavior**
When I navigate to the app the error I see is:
Internal Server Error
The server was unable to complete your request.
If this happens again, please send the technical details below to the server administrator.
More details can be found in the server log.

Looking for log file has been difficult:
/snap/nextcloud/current/logs   is empty
Searching for the term 'log' in /snap/nextcloud/current brought up too many options to go through, I tried a few more promising named files/folders and was left confused.

**Screenshots**

**Browser**
Firefox 88

**Versions**
Nextcloud server version: Using snap on Ubuntu 
Cookbook version: 7.6
Database system: using the default snap installation/unsure

This is all my main problem but in addition clicking to update the app to 8.4 in Nextcloud visually looks like it is working but reloading the page reveals nothing has happened. The same goes for trying to remove the app.

**Comment by christianlupus** [comment]
Unfortunately, it is hard to debug the exact problem.

I suspect you mean cookbook version 0.7.6 and 0.8.4, right?
The 5xx error you mentioned is visible in the browser, correct?

Is there anything logged within the nextcloud logs?

It sounds like you have a problem with your installation. You could have a look at the developer tools if there are any suspicious requests/answers from the server.

Currently, you can easily remove the app and keep the file structure of your files in a backup. After reinstalling the current version, you can just copy the files to the recipes' folder and trigger a database renewal.
If removal of the app is impossible this is a real indicator that something suspicious has happened to the installation. Can you access the database directly? With some luck, you could remove the app manually but it might as well be only a symptom of another issue.

**Comment by Boballen55** [comment]
I ended up down a confusing network issue for a while that I think I've straightened out now, I thought possibly it was related to this issue but this still persists. I'm using the snap install and I struggle to figure out where the relevant log files are on my system, where should I be looking?

What do I need to remove from where to "remove" the app without breaking the database. I thought this was what should be happening if I clicked remove in the Nextcloud app tab. I could try guess and checking some folders but the rest of my instance isn't giving me issues.

**Comment by IzzySoft** [comment]
I see a similar behavior here with 0.8.2 (yes, I just reached NC17 on my test instance now that I finally have PHP > 7.0 available). For me, the logs state `class +'OCP\DB\Types' not found` (triggered from `cookbook/lib/Db/DbTypesPolyfillHelper.php`). Could this be related? Besides, if I fix that by assigning the 2 types directly, the logs instead show `Could not resolve recipeService! Class recipeService does not exist` (so it obviously did not like `$this->int = 'integer';`).

Work around for me was downgrading to 0.7.6 again (replacing the `apps/cookbook` directory manually and running `update oc_appconfig set configvalue='0.7.6' where appid='cookbook' and configkey='installed_version';` against the database).

@Boballen55 what PHP version is your snap using? You might need to downgrade even further. Had that on another installation where the machine had only PHP7.0 (and Nextcloud 15, as that's the latest running with 7.0). The latest cookbook version working with that would be 0.6.3, if you wanted to try that. IIRC 0.7.6 already requires PHP7.2 (0.8.6 definitely does).

**Comment by christianlupus** [comment]
> I see a similar behavior here with 0.8.2 (yes, I just reached NC17 on my test instance now that I finally have PHP > 7.0 available).

I hope, you are aware, that NC17 is past its end of life quite for some time now?

> For me, the logs state `class +'OCP\DB\Types' not found` (triggered from `cookbook/lib/Db/DbTypesPolyfillHelper.php`). Could this be related?

No, this is unrelated. The polyfill helper is designed only for the supported versions 18 and later. One could make it compatible with the older versions of the core NC but I hoped we could get rid of the polyfill rather sooner than later.

> Besides, if I fix that by assigning the 2 types directly, the logs instead show `Could not resolve recipeService! Class recipeService does not exist` (so it obviously did not like `$this->int = 'integer';`).

I think this is some problem with an old PHP version. I stumbled upon a similar issue but do not remember if this is a pre-7.1, 7.2 or 7.3 problem. I did not find it just at the moment.

> Work around for me was downgrading to 0.7.6 again (replacing the `apps/cookbook` directory manually and running `update oc_appconfig set configvalue='0.7.6' where appid='cookbook' and configkey='installed_version';` against the database).

Keep in mind that you need to keep the database schema compatible with the installed app version. Otherwise you might get *interesting* effects.

> @Boballen55 what PHP version is your snap using? You might need to downgrade even further. Had that on another installation where the machine had only PHP7.0 (and Nextcloud 15, as that's the latest running with 7.0). The latest cookbook version working with that would be 0.6.3, if you wanted to try that. IIRC 0.7.6 already requires PHP7.2 (0.8.6 definitely does).

You mean 0.8.4 not 0.8.6, right? We need even PHP 7.3 currently.

Do you use [the snap package build from this source](https://github.com/nextcloud/nextcloud-snap)? You can have a look [here](https://github.com/nextcloud/nextcloud-snap#debug-mode) and [here](https://github.com/nextcloud/nextcloud-snap#where-is-my-stuff).
Is there anything logged in the nextcloud logs (I mean not the apache/PHP logs)?

**Comment by IzzySoft** [comment]
> I hope, you are aware, that NC17 is past its end of life quite for some time now?

Yes, I am. And I'll certainly move up to NC18 soon – but then will be stuck again as NC19 (or was it NC20) already requires PHP7.3+, while my system only comes with 7.2. I prefer a step-by-step upgrade, to fix issues one by one. Was stuck on NC15 for much too long, waiting for PHP > 7.0 being available for my system. I now have 7.2, so choices are still limited.

Nevertheless: 0.8.2 was indicated to work with NC17, but unfortunately is not. I don't expect anyone to fix *that* – I just wanted to offer a work-around for those who for some reason cannot upgrade further.

> No, this is unrelated. The polyfill helper is designed only for the supported versions 18 and later.

Well, then 0.8.2 should not indicate minVer 17 but 18 – and thus wouldn't be pulled in when updating to NC 17.

> Keep in mind that you need to keep the database schema compatible with the installed app version.

I'm aware of that. It worked out for me this way, otherwise I had pulled the corresponding tables from the backup.

For the remaining part: I'm not using snap at all (I'd never do that), I use the distributed tarball and install it manually. What I found in the logs I included with my post above. For now, I'm happy using 0.7.6 until I can move on to higher grounds; as pointed out, I was just chiming in to share the work-around that worked for me.

**Comment by IzzySoft** [comment]
Just a short update: issue remains on NC18 (I almost was expecting this, as the cookbook version is the very same). The very same remedy works (downgrading to 0.7.6). Curious what might happen on NC19 then, as that will permit a newer cookbook version then (0.8.4 instead of 0.8.2). I hope it's solved there, as with NC20 we can no longer revert to 0.7.2 (which has `max-version="19"`).

**Comment by IzzySoft** [comment]
Next update with good news: Having reached Nextcloud 19 (the oldest still supported version) upgraded Cookbook to 0.8.4 – and it works! :tada: :rocket:

Now the only problem for me is: that was my test instance, running on Debian Buster, which comes with PHP7.3. My main instance only has PHP7.2 available – so if I want to use Cookbook, I cannot upgrade it to 0.8.4 as that requires PHP7.3. But good to know that the issue is solved; for now, I can upgrade to NC19 (which supports PHP7.2) and stick to Cookbook 0.7.6 (which supports the combination) – but then I'll be stuck as while NC20 supports PHP7.2, Cookbook 0.7.6 does not support NC20…

Before you ask: yes, I'm running an LTS version supported until 2023. And it's not always possible to "just update" to the latest stuff – that's why LTS versions have 5 years support, so you're not sitting on a constantly moving target :wink: So I'm not complaining: I'll stay a happy Cookbook user, so thanks a lot for this great app! I just wish apps would also keep LTS versions in mind before switching to the latest libs. Well, can't have it all. TL;DR: heads up, and keep the book cooking please :smiley:

**Comment by christianlupus** [comment]
OK, this issue has gone out of shape.

Regarding the original topic (snap-based), I have not seen much progress in the discussion. I will enable a stale bot so this issue will eventually be closed when left alone.

@IzzySoft I think your problem is mainly the PHP 7.2 support. It might be best to open a new issue there. It might be an option to tweak your installation to support such old versions **but** be aware that NC 19 is already EOL and NC20 will be [in mere months](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule). Also, PHP 7.2 is [no longer supported](https://www.php.net/supported-versions.php) independent of any LTS considerations (or better the code PHP does not have an LTS support as far as I can see).

**Comment by IzzySoft** [comment]
@christianlupus ***I*** am aware. But what is demotivating is that LTS versions of distributions are that disregarded. What good is 5y support if one must upgrade after just 1.5 years due to software running out of maintenance (it's not just NC/Cookbook)? The one running on my desktop currently is supported until 4/2023 but will never go beyond PHP7.2 – so the highest version of NC I can run on that will be 20; without Cookbook support then, as there's no version compatible with both NC20 and PHP7.2, so in reality it's rather NC19.

That said: NC19 on Buster (PHP7.3) runs fine with its corresponding Cookbook version – which hopefully holds true for newer versions as well, so my issue is indeed "historical" (no need for a spin-off as far as I'm concerned). I hope I can upgrade my main box from Mint 19.3 to 20 soon.

**TL;DR:** Feel free to have this issue "run stale", looks like I can help myself around (and those having no plan how to do so hopefully can fish the work-arounds from above comments). Thanks!

**Comment by Boballen55** [comment]
> Do you use [the snap package build from this source](https://github.com/nextcloud/nextcloud-snap)? You can have a look [here](https://github.com/nextcloud/nextcloud-snap#debug-mode) and [here](https://github.com/nextcloud/nextcloud-snap#where-is-my-stuff).
> Is there anything logged in the nextcloud logs (I mean not the apache/PHP logs)?

I am using the snap from that source and I am at version 21.0.3snap1. I've attempted to turn on the debug mode to create additional logs but from the link it looks like that will only give more detailed Apache/PHP logs. I did find a bunch of logs under /var/snap/nextcloud/current/logs. However, I don't see a nextcloud.log. I did try using, then updating, then disabling and removing the cookbook app again. I did not see anything that looked releated in the any of the logs I did find (Apache, PHP, MySQL, Redis). The nextcloud forums seem to be struggling right now

**Comment by christianlupus** [comment]
@Boballen55 I am unsure how snap works internally. However, there should be a folder with the user data. Within this folder, there should be one folder per NC user containing the `files` etc. Next to the folders for the users, there should be a `nextcloud.log`, which might be of interest.

At least currently, the forum/discourse of NC is working. Maybe they can help you where to find the logs. Without them, it is merely plain guessing.
